arangod/Replication2/ReplicatedLog/AgencyLogSpecification.h

```diff
@@ -0,0 +1,131 @@
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Lars Maier
+////////////////////////////////////////////////////////////////////////////////
+#pragma once
+
+#include "Agency/AgencyPaths.h"
+#include "Cluster/ClusterTypes.h"
+#include "Replication2/ReplicatedLog/LogCommon.h"
+#include "Replication2/ReplicatedLog/types.h"
+
+#include <velocypack/Builder.h>
+#include <velocypack/Slice.h>
+
+#include <optional>
+
+namespace arangodb::replication2::agency {
+
+struct from_velocypack_t {};
+inline constexpr auto from_velocypack = from_velocypack_t{};
+
+struct LogPlanTermSpecification {
+  LogTerm term;
+  LogConfig config;
+  struct Leader {
+    ParticipantId serverId;
+    RebootId rebootId;
+  };
+  std::optional<Leader> leader;
+
+  struct Participant {};
+  std::unordered_map<ParticipantId, Participant> participants;
+
+  auto toVelocyPack(VPackBuilder&) const -> void;
+  LogPlanTermSpecification(from_velocypack_t, VPackSlice);
+  LogPlanTermSpecification() = default;
+
+  LogPlanTermSpecification(LogTerm term, LogConfig config, std::optional<Leader>, std::unordered_map<ParticipantId, Participant> participants);
+};
+
+struct LogPlanSpecification {
+  LogId id;
+  std::optional<LogPlanTermSpecification> currentTerm;
+
+  LogConfig targetConfig;
+
+  auto toVelocyPack(VPackBuilder&) const -> void;
+  LogPlanSpecification(from_velocypack_t, VPackSlice);
+  LogPlanSpecification() = default;
+
+  LogPlanSpecification(LogId id, std::optional<LogPlanTermSpecification> term, LogConfig config);
+};
+
+struct LogCurrentLocalState {
+  LogTerm term{};
+  TermIndexPair spearhead{};
+
+  auto toVelocyPack(VPackBuilder&) const -> void;
+  LogCurrentLocalState() = default;
+  LogCurrentLocalState(from_velocypack_t, VPackSlice);
+  LogCurrentLocalState(LogTerm, TermIndexPair) noexcept;
+};
+
+
+struct LogCurrentSupervisionElection {
+  enum class ErrorCode {
+    OK = 0,
+    SERVER_NOT_GOOD = 1,
+    TERM_NOT_CONFIRMED = 2,
+  };
+
+  LogTerm term;
+  std::size_t participantsRequired{};
+  std::size_t participantsAvailable{};
+  std::unordered_map<ParticipantId, ErrorCode> detail;
+
+  auto toVelocyPack(VPackBuilder&) const -> void;
+
+  friend auto operator==(LogCurrentSupervisionElection const&,
+                         LogCurrentSupervisionElection const&) noexcept -> bool;
+  friend auto operator!=(LogCurrentSupervisionElection const& left,
+                         LogCurrentSupervisionElection const& right) noexcept -> bool {
+    return !(left == right);
+  }
+
+  LogCurrentSupervisionElection() = default;
+  LogCurrentSupervisionElection(from_velocypack_t, VPackSlice slice);
+};
+
+auto operator==(LogCurrentSupervisionElection const&,
+                LogCurrentSupervisionElection const&) noexcept -> bool;
```

Comment:
This declaration is not redundant because of 

> 7.3.1.2.3 Every name first declared in a namespace is a member of that namespace. If a friend declaration in a non-local class first declares a class, function, class template or function template the friend is a member of the innermost enclosing namespace. The friend declaration does not by itself make the name visible to unqualified lookup or qualified lookup.

However, when the operator is used as operator, then it is looked up using [ADL](https://en.cppreference.com/w/cpp/language/adl), which explicitly states that:
> The set of declarations found by ordinary unqualified lookup and the set of declarations found in all elements of the associated set produced by ADL, are merged, with the following special rules
> 1) using-directives in the associated namespaces are ignored
> 2) namespace-scoped friend functions (and function templates) that are declared in an associated class are visible through ADL even if they are not visible through ordinary lookup
> 3) all names except for the functions and function templates are ignored (no collision with variables)

[This](https://godbolt.org/z/6h3MvzeET) example shows that a forward declaration is required.

Reply:
True, the inline friend declaration is only accessible via argument-dependent-lookup (ADL), but since an operator is usually only ever used with actual arguments, I would argue that this is good enough. However, I have no strong opinion here, so if you prefer just leave it as is.